### DISCUSSION TITLE
Request for feedback: Rspack-based bundler builder

### DISCUSSION THREAD

**Discussion by h-a-n-a** [web-infra-dev/rspack#4585] [Ideas]
## TL;DR

Rspack-based bundler builder integrates user-defined rust plugins and loaders while still having the ability for Webpack API interoperability.

## What is it?

Just like how Node did to V8, they are integrated by including V8's source code in their project. For Rspack, to provide better integration with infrastructure in various scenario, Rspack-based bundler builder provides everything that we have right now. Things like Webpack Loader compatible API and Plugin compatible API are included, and stable Node API will be included soon. These feature are stable. Moreover, a bundler builder makes things like **writing Rust plugin** possible. **User-defined rust plugins will be compiled together with rspack core while still having the ability for Webpack API interoperability.**

## Possible architecture

![whiteboard_exported_image (1)](https://github.com/web-infra-dev/rspack/assets/10465670/3f8c7c88-4f86-4f60-80c7-113fc38f7e93)

### Cornerstone: Rust crate

In the current version of Rspack, we does not guarantee the stability for the Rust crate, which results in people can barely rely on this and build something on top of Rspack crate. In order to make rspack-based bundler builder working. Rspack rust crate will **provide stable Rust plugin and loader interface** without losing the performance when creating WASM-based runtime to ensure the best overall performance.

**This includes:**

- Stability for rspack rust core

### Bundler Builder:

Builder builder is built on the stability of the cornerstone. In order to maximize the capability of developer designing a bundler in one company, we would love to provide best-effort out-of-the-box support for debugging and releasing.

**This includes:**

- Bundler builder adapter layer to integrate user-defined Rspack Rust plugins and loaders.
- A CLI for helping our day-to-day development.
- Bundler builder core for supporting generating JS binding code and GitHub Action CI/Testing workflow (This does not guarantee to be stable).
- Debug symbol management for testing and debugging
- ...

### Artifacts and publishing

The generated artifacts for Builder builder is basically what we have today, platform-based node addons with user-defined platform to which be targeting to, a JS core providing the JS interoperability.
